File: java/MarkDuplicates.java
/*
	Copyright (c) 2018 by imec vzw, Leuven, Belgium. All rights reserverd.
*/

package elprep;

import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.*;
import java.util.function.*;

public class MarkDuplicates {

  static class Fragment {
    Slice lb;
    int refid;
    int pos;
    boolean reversed;

    Fragment (Slice lb, int refid, int pos, boolean reversed) {
      this.lb = lb;
      this.refid = refid;
      this.pos = pos;
      this.reversed = reversed;
    }

    public boolean equals (Object obj) {
      var that = (Fragment)obj;
      return this.lb.equals(that.lb) &&
        (this.refid == that.refid) &&
        (this.pos == that.pos) &&
        (this.reversed == that.reversed);
    }

    public int hashCode () {
      return lb.hashCode() ^ Integer.hashCode(refid) ^ Integer.hashCode(pos) ^ Boolean.hashCode(reversed);
    }
  }

  static void classifyFragment (SamAlignment aln, ConcurrentMap<Fragment, AtomicReference<SamAlignment>> fragments, boolean deterministic) {
    var key = new Fragment(aln.getLIBID(), aln.getREFID(), aln.getAdaptedPos(), aln.isReversed());
    var best = fragments.putIfAbsent(key, new AtomicReference<>(aln));
    if (best != null) {
      if (aln.isTrueFragment()) {
        var alnScore = aln.getAdaptedScore();
        while (true) {
          var bestAln = best.get();
          if (bestAln.isTruePair()) {
            aln.FLAG |= SamAlignment.Duplicate; break;
          } else {
            var bestAlnScore = bestAln.getAdaptedScore();
            if (bestAlnScore > alnScore) {
              aln.FLAG |= SamAlignment.Duplicate; break;
            } else if (bestAlnScore == alnScore) {
              if (deterministic) {
                if (aln.QNAME.compareTo(bestAln.QNAME) > 0) {
                  aln.FLAG |= SamAlignment.Duplicate; break;
                } else if (best.compareAndSet(bestAln, aln)) {
                  bestAln.FLAG |= SamAlignment.Duplicate; break;
                }
              } else {
                aln.FLAG |= SamAlignment.Duplicate; break;
              }
            } else if (best.compareAndSet(bestAln, aln)) {
              bestAln.FLAG |= SamAlignment.Duplicate; break;
            }
          }
        }
      } else {
        while (true) {
          var bestAln = best.get();
          if (bestAln.isTruePair()) {
            break;
          } else if (best.compareAndSet(bestAln, aln)) {
            bestAln.FLAG |= SamAlignment.Duplicate; break;
          }
        }
      }
    }
  }

  static class PairFragment {
    Slice lb;
    Slice qname;

    PairFragment (Slice lb, Slice qname) {
      this.lb = lb;
      this.qname = qname;
    }

    public boolean equals (Object obj) {
      var that = (PairFragment)obj;
      return this.lb.equals(that.lb) &&
        this.qname.equals(that.qname);
    }

    public int hashCode () {
      return lb.hashCode() ^ qname.hashCode();
    }
  }

  static class Pair {
    Slice lb;
    int refid1, refid2;
    int pos1, pos2;
    boolean reversed1, reversed2;

    Pair (Slice lb, int refid1, int refid2, int pos1, int pos2, boolean reversed1, boolean reversed2) {
      this.lb = lb;
      this.refid1 = refid1;
      this.refid2 = refid2;
      this.pos1 = pos1;
      this.pos2 = pos2;
      this.reversed1 = reversed1;
      this.reversed2 = reversed2;
    }

    public boolean equals (Object obj) {
      var that = (Pair)obj;
      return this.lb.equals(that.lb) &&
        (this.refid1 == that.refid1) &&
        (this.refid2 == that.refid2) &&
        (this.pos1 == that.pos1) &&
        (this.pos2 == that.pos2) &&
        (this.reversed1 == that.reversed1) &&
        (this.reversed2 == that.reversed2);
    }

    public int hashCode () {
      return lb.hashCode() ^
        Integer.hashCode(refid1) ^
        Integer.hashCode(refid2) ^
        Long.hashCode((((long)pos1) << 23) ^ (long)pos2) ^
        Boolean.hashCode(reversed1) ^
        Boolean.hashCode(reversed2);
    }
  }

  static class SamAlignmentPair {
    int score;
    SamAlignment aln1, aln2;

    SamAlignmentPair (int score, SamAlignment aln1, SamAlignment aln2) {
      this.score = score;
      this.aln1 = aln1;
      this.aln2 = aln2;
    }
  }

  static void classifyPair
    (SamAlignment aln,
     ConcurrentMap<PairFragment, SamAlignment> fragments,
     ConcurrentMap<Pair, AtomicReference<SamAlignmentPair>> pairs,
     boolean deterministic) {
    if (!aln.isTruePair()) {return;}

    var aln1 = aln;
    var fragmentKey = new PairFragment(aln.getLIBID(), aln.QNAME);
    var aln2 = fragments.putIfAbsent(fragmentKey, aln);
    if ((aln2 != null) && fragments.remove(fragmentKey, aln2)) {

      var score = aln1.getAdaptedScore() + aln2.getAdaptedScore();
      var aln1Pos = aln1.getAdaptedPos();
      var aln2Pos = aln2.getAdaptedPos();
      if (aln1Pos > aln2Pos) {
        var alnTemp = aln1;
        aln1 = aln2;
        aln2 = alnTemp;
        var temp = aln1Pos;
        aln1Pos = aln2Pos;
        aln2Pos = temp;
      }

      var pairKey = new Pair
        (aln1.getLIBID(),
         aln1.getREFID(),
         aln2.getREFID(),
         aln1Pos,
         aln2Pos,
         aln1.isReversed(),
         aln2.isReversed());

      var newPair = new SamAlignmentPair(score, aln1, aln2);

      var best = pairs.putIfAbsent(pairKey, new AtomicReference<>(newPair));
      if (best != null) {
        while (true) {
          var bestPair = best.get();
          if (bestPair.score > score) {
            aln1.FLAG |= SamAlignment.Duplicate;
            aln2.FLAG |= SamAlignment.Duplicate;
            break;
          } else if (bestPair.score == score) {
            if (deterministic) {
              if (aln1.QNAME.compareTo(bestPair.aln1.QNAME) > 0) {
                aln1.FLAG |= SamAlignment.Duplicate;
                aln2.FLAG |= SamAlignment.Duplicate;
                break;
              } else if (best.compareAndSet(bestPair, newPair)) {
                bestPair.aln1.FLAG |= SamAlignment.Duplicate;
                bestPair.aln2.FLAG |= SamAlignment.Duplicate;
                break;
              }
            } else {
              aln1.FLAG |= SamAlignment.Duplicate;
              aln2.FLAG |= SamAlignment.Duplicate;
              break;
            }
          } else if (best.compareAndSet(bestPair, newPair)) {
            bestPair.aln1.FLAG |= SamAlignment.Duplicate;
            bestPair.aln2.FLAG |= SamAlignment.Duplicate;
            break;
          }
        }
      }
    }
  }

  private static Slice LB = new Slice("LB");
  private static Slice ID = new Slice("ID");

  public static Function<SamHeader, Predicate<SamAlignment>> markDuplicates (boolean deterministic) {
    return (header) -> {
      var fragments = new ConcurrentHashMap<Fragment, AtomicReference<SamAlignment>>();
      var pairsFragments = new ConcurrentHashMap<PairFragment, SamAlignment>();
      var pairs = new ConcurrentHashMap<Pair, AtomicReference<SamAlignmentPair>>();
      var lbTable = new HashMap<Slice, Slice>();
      for (var rgEntry: header.RG) {
        var lb = rgEntry.get(LB);
        if (lb != null) {
          var id = rgEntry.get(ID);
          if (id == null) {
            throw new RuntimeException("Missing mandatory ID entry in an @RG line in a SAM file header.");
          }
          lbTable.put(id, lb);
        }
      }
      return (aln) -> {
        if (aln.flagNotAny(SamAlignment.Unmapped |
                           SamAlignment.Secondary |
                           SamAlignment.Duplicate |
                           SamAlignment.Supplementary)) {
          aln.adaptAlignment(lbTable);
          classifyFragment(aln, fragments, deterministic);
          classifyPair(aln, pairsFragments, pairs, deterministic);
        }
        return true;
      };
    };
  }
}

File: cpp/mark_duplicates.cpp
/*
	Copyright (c) 2018 by imec vzw, Leuven, Belgium. All rights reserverd.
*/

const auto phred_score_table = []() -> array<uint8_t,512> {
	array<uint8_t,512> table;
	for (auto c = 0; c < 256; ++c) {
		auto pos = c << 1;
		if ((c < 33) || (c > 126)) {
			table[pos] = 0;
			table[pos+1] = 1;
		} else {
			auto qual = c - 33;
			table[pos] = (qual >= 15) ? qual : 0;
			table[pos+1] = 0;
		}
	}
	return table;
}();

int32_t compute_phred_score (const shared_ptr<sam_alignment>& aln) {
	int32_t score = 0;
	int32_t error = 0;
	for (auto p = aln->qual.begin(); p != aln->qual.end(); ++p) {
		auto pos = *p << 1;
		score += phred_score_table[pos];
		error |= phred_score_table[pos+1];
	}
	if (error != 0) {
		throw runtime_error("Invalid QUAL character.");
	}
	return score;
}

const set<char> clipped_table({'S', 'H'});
const set<char> reference_table({'M', 'D', 'N', '=', 'X'});

int32_t compute_unclipped_position(const shared_ptr<sam_alignment>& aln) {
	auto& cigar = scan_cigar_string(aln->cigar);
	if (cigar.size() == 0) {return aln->pos;}

	if (aln->is_reversed()) {
		int32_t clipped = 1;
		auto result = aln->pos - 1;
		for (int i = cigar.size()-1; i >= 0; --i) {
			auto op = cigar[i];
			auto p = op.operation;
			auto c = clipped_table.count(p);
			auto r = reference_table.count(p);
			clipped *= c;
			result += (r|clipped)*op.length;
		}
		return result;
	} else {
		auto result = aln->pos;
		for (auto op: cigar) {
			auto p = op.operation;
			if (clipped_table.count(p) == 0) {break;}
			result -= op.length;
		}
		return result;
	}
}

const string_slice pos("pos");
const string_slice score("score");

inline int32_t get_adapted_pos (const shared_ptr<sam_alignment>& aln) {
	return any_cast<int32_t>(assoc(aln->temps, pos)->value);
}

inline void set_adapted_pos (const shared_ptr<sam_alignment>& aln, int32_t value) {
	auto it = assoc(aln->temps, pos);
	if (it == aln->temps.end()) {
		aln->temps.push_back(sam_value(pos, value));
	} else {
		it->value = value;
	}
}

inline int32_t get_adapted_score (const shared_ptr<sam_alignment>& aln) {
	return any_cast<int32_t>(assoc(aln->temps, score)->value);
}

inline void set_adapted_score (const shared_ptr<sam_alignment>& aln, int32_t value) {
	auto it = assoc(aln->temps, score);
	if (it == aln->temps.end()) {
		aln->temps.push_back(sam_value(score, value));
	} else {
		it->value = value;
	}
}

void adapt_alignment (const shared_ptr<sam_alignment>& aln, const string_map& lb_table) {
	auto rg = aln->get_rg();
	auto it = lb_table.find(rg);
	if (it != lb_table.end()) {
		aln->set_libid(it->second);
	}
	set_adapted_pos(aln, compute_unclipped_position(aln));
	set_adapted_score(aln, compute_phred_score(aln));
}

template<typename T>
class handle {
public:
	T object;

	handle (const T& object) : object(object) {}

	bool compare_exchange (T& expected, const T& desired) {
		return atomic_compare_exchange_strong(&object, &expected, desired);
	}
};

inline bool is_true_fragment (const shared_ptr<sam_alignment>& aln) {
	return (aln->flag & (multiple | next_unmapped)) != multiple;
}

inline bool is_true_pair (const shared_ptr<sam_alignment>& aln) {
	return (aln->flag & (multiple | next_unmapped)) == multiple;
}

class fragment_hash {
public:
	size_t operator() (const shared_ptr<sam_alignment>& aln) const {
		return 29 * (tbb_hasher(aln->get_libid()) ^
								 tbb_hasher(aln->get_refid()) ^
								 tbb_hasher(get_adapted_pos(aln)) ^
								 tbb_hasher(aln->is_reversed()));
	}
};

class fragment_equal {
public:
	bool operator() (const shared_ptr<sam_alignment>& aln1, const shared_ptr<sam_alignment>& aln2) const {
		return (aln1->get_libid() == aln2->get_libid()) &&
			(aln1->get_refid() == aln2->get_refid()) &&
			(get_adapted_pos(aln1) == get_adapted_pos(aln2)) &&
			(aln1->is_reversed() == aln2->is_reversed());
	}
};

typedef concurrent_unordered_map<shared_ptr<sam_alignment>, shared_ptr<handle<shared_ptr<sam_alignment>>>, fragment_hash, fragment_equal> fragment_map;

void classify_fragment (const shared_ptr<sam_alignment>& aln, fragment_map& fragments, bool deterministic) {
	auto p = fragments.emplace(aln, make_shared<handle<shared_ptr<sam_alignment>>>(aln));
	if (p.second) return;
	auto best_handle = p.first->second;
	if (is_true_fragment(aln)) {
		auto aln_score = get_adapted_score(aln);
		while (true) {
			auto best = best_handle->object;
			if (is_true_pair(best)) {
				aln->flag |= duplicate; break;
			} else {
				auto best_aln_score = get_adapted_score(best);
				if (best_aln_score > aln_score) {
					aln->flag |= duplicate; break;
				} else if (best_aln_score == aln_score) {
					if (deterministic) {
						if (aln->qname > best->qname) {
							aln->flag |= duplicate; break;
						} else if (best_handle->compare_exchange(best, aln)) {
							best->flag |= duplicate; break;
						}
					} else {
						aln->flag |= duplicate; break;
					}
				} else if (best_handle->compare_exchange(best, aln)) {
					best->flag |= duplicate; break;
				}
			}
		}
	} else {
		while (true) {
			auto best = best_handle->object;
			if (is_true_pair(best)) {
				break;
			} else if (best_handle->compare_exchange(best, aln)) {
				best->flag |= duplicate; break;
			}
		}
	}
}

class alignment_pair_hash {
public:
	static size_t hash (const shared_ptr<sam_alignment>& aln) {
		return 29 * (tbb_hasher(aln->get_libid()) ^ tbb_hasher(aln->qname));
	}

	static bool equal (const shared_ptr<sam_alignment>& aln1, const shared_ptr<sam_alignment>& aln2) {
		return (aln1->get_libid() == aln2->get_libid()) && (aln1->qname == aln2->qname);
	}
};

typedef concurrent_hash_map<shared_ptr<sam_alignment>, shared_ptr<sam_alignment>, alignment_pair_hash> pair_fragment_map;

class pair_handle {
public:
	int32_t score;
	shared_ptr<sam_alignment> aln1;
	shared_ptr<sam_alignment> aln2;

	pair_handle (int32_t score, const shared_ptr<sam_alignment>& aln1, const shared_ptr<sam_alignment>& aln2) : score(score), aln1(aln1), aln2(aln2) {}
	pair_handle (const pair_handle& p) : score(p.score), aln1(p.aln1), aln2(p.aln2) {}
};

class pair_hash {
public:
	size_t operator() (const pair_handle& p) const {
		return 29 * (tbb_hasher(p.aln1->get_libid()) ^
								 tbb_hasher(p.aln1->get_refid()) ^
								 tbb_hasher(p.aln2->get_refid()) ^
								 tbb_hasher((int64_t(get_adapted_pos(p.aln1)) << 32) + int64_t(get_adapted_pos(p.aln2))) ^
								 tbb_hasher(p.aln1->is_reversed()) ^
								 tbb_hasher(p.aln2->is_reversed()));
	}
};

class pair_equal {
public:
	bool operator() (const pair_handle& p1, const pair_handle& p2) const {
		return (p1.aln1->get_libid() == p2.aln1->get_libid()) &&
			(p1.aln1->get_refid() == p2.aln1->get_refid()) &&
			(get_adapted_pos(p1.aln1) == get_adapted_pos(p2.aln1)) &&
			(p1.aln1->is_reversed() == p2.aln1->is_reversed()) &&
			(p1.aln2->get_refid() == p2.aln2->get_refid()) &&
			(get_adapted_pos(p1.aln2) == get_adapted_pos(p2.aln2)) &&
			(p1.aln2->is_reversed() == p2.aln2->is_reversed());
	}
};

typedef concurrent_unordered_map<pair_handle, shared_ptr<handle<shared_ptr<pair_handle>>>, pair_hash, pair_equal> pair_map;

void classify_pair (const shared_ptr<sam_alignment>& aln, pair_fragment_map& fragments, pair_map& pairs, bool deterministic) {
	if (!is_true_pair(aln)) return;

	auto aln1 = aln;
	auto aln2 = [&]() -> shared_ptr<sam_alignment> {
		pair_fragment_map::accessor acc;
		if (fragments.insert(acc, aln)) {
			acc->second = aln;
			return nullptr;
		} else {
			auto aln2 = acc->second;
			fragments.erase(acc);
			return aln2;
		}
	}();

	if (aln2 == nullptr) return;

	auto score = get_adapted_score(aln1) + get_adapted_score(aln2);
	auto aln1pos = get_adapted_pos(aln1);
	auto aln2pos = get_adapted_pos(aln2);
	if (aln1pos > aln2pos) {
		swap(aln1, aln2);
		swap(aln1pos, aln2pos);
	}

	pair_handle pair_key(score, aln1, aln2);

	auto p = pairs.emplace(pair_key, make_shared<handle<shared_ptr<pair_handle>>>(make_shared<pair_handle>(pair_key)));
	if (p.second) return;
	auto best_handle = p.first->second;
	shared_ptr<pair_handle> entry = nullptr;
	auto get_entry = [&](){
		if (entry == nullptr) {
			entry = make_shared<pair_handle>(pair_key);
		}
		return entry;
	};
	while (true) {
		auto best = best_handle->object;
		if (best->score > score) {
			aln1->flag |= duplicate;
			aln2->flag |= duplicate;
			break;
		} else if (best->score == score) {
			if (deterministic) {
				if (aln1->qname > best->aln1->qname) {
					aln1->flag |= duplicate;
					aln2->flag |= duplicate;
					break;
				} else if (best_handle->compare_exchange(best, get_entry())) {
					best->aln1->flag |= duplicate;
					best->aln2->flag |= duplicate;
					break;
				}
			} else {
				aln1->flag |= duplicate;
				aln2->flag |= duplicate;
				break;
			}
		} else if (best_handle->compare_exchange(best, get_entry())) {
			best->aln1->flag |= duplicate;
			best->aln2->flag |= duplicate;
			break;
		}
	}
}

header_filter mark_duplicates (bool deterministic) {
	return [deterministic](const shared_ptr<sam_header>& header) -> alignment_filter {
		auto fragments = make_shared<fragment_map>(1000000);
		auto pair_fragments = make_shared<pair_fragment_map>(1000000);
		auto pairs = make_shared<pair_map>(1000000);
		string_map lb_table;
		for (auto& rg_entry: header->rg) {
			auto lb_it = rg_entry.find(LB);
			if (lb_it != rg_entry.end()) {
				auto id_it = rg_entry.find(ID);
				if (id_it == rg_entry.end()) {
					throw runtime_error("Missing mandatory ID entry in an @RG line in a SAM file header.");
				}
				lb_table.emplace(id_it->second, lb_it->second);
			}
		}
		return [lb_table, fragments, pair_fragments, pairs, deterministic](const shared_ptr<sam_alignment>& aln) -> bool {
			if (aln->flag_not_any(unmapped | secondary | duplicate | supplementary)) {
				adapt_alignment(aln, lb_table);
				classify_fragment(aln, *fragments, deterministic);
				classify_pair(aln, *pair_fragments, *pairs, deterministic);
			}
			return true;
		};
	};
}
